### ISSUE TITLE
Tracking issue for `Option::contains` and `Result::contains`

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#62358]
This is a tracking issue for `Option::contains` and `Result::contains`.

The implementations are as follows:
```rust
    pub fn contains(&self, x: &T) -> bool where T: PartialEq {
        match self {
            Some(y) => y == x,
            None => false,
        }
    }

    pub fn contains(&self, x: &T) -> bool where T: PartialEq {
        match self {
            Ok(y) => y == x,
            Err(_) => false,
        }
    }
``

**Comment by Mark-Simulacrum** [comment]
Should we update the signatures to `fn contains<U>(&self, x: &U) -> bool where T: PartialEq<U>`?

**Comment by Centril** [comment]
@Mark-Simulacrum That seems like a good idea -- type inference should be driven by `x` in any case. Let's do that in the PR.

**Comment by timvermeulen** [comment]
Looks like `slice`/`Vec`/`VecDeque`/`LinkedList` and possibly others have a `contains` method that could benefit in the same way.

**Comment by soc** [comment]
@Mark-Simulacrum I didn't do that because I checked the collections first, and they were all invariant. Shall I change this for `Option` and `Result` only?

**Comment by Mark-Simulacrum** [comment]
We usually can't change preexisting APIs due to inference breakage, but since this is new, we should be able to be general.

**Comment by frewsxcv** [comment]
Are these functionally equivalent?

```rust
    pub fn contains(&self, x: &T) -> bool where T: PartialEq {
        match self {
            Some(y) => y == x,
            None => false,
        }
    }
```

```rust
    pub fn contains(&self, x: &T) -> bool where T: PartialEq {
        self.as_ref() == Some(x)
    }
```

**Comment by soc** [comment]
Not quite, due to the requested variance changes.

**Comment by czipperz** [comment]
What does that mean @soc?

**Comment by soc** [comment]
@czipperz The implementation is:

    // Option
    pub fn contains<U>(&self, x: &U) -> bool where U: PartialEq<T> {
        match self {
            Some(y) => x == y,
            None => false,
        }
    }

    // Result
    pub fn contains<U>(&self, x: &U) -> bool where U: PartialEq<T> {
        match self {
            Ok(y) => x == y,
            Err(_) => false
        }
    }

    pub fn contains_err<F>(&self, f: &F) -> bool where F: PartialEq<E> {
        match self {
            Ok(_) => false,
            Err(e) => f == e
        }
    }

This means that the argument doesn't have to be of the same type as the value inside option/result, only that it knows how to compare itself to it.

**Comment by czipperz** [comment]
So there's no implementation of `PartialEq<Option<U>> for Option<T> where U: PartialEq<T>`?

**Comment by timvermeulen** [comment]
There is indeed not, see https://github.com/rust-lang/rfcs/issues/1239.

**Comment by dtolnay** [comment]
These methods are way to high up in the docs on both the `Option` and `Result` rustdoc. Would someone be able to send a PR to move these a lot further down? For example stuff like `ok` and `map` and `and_then` should all be featured higher than contains.

**Comment by sinistersnare** [comment]
What is the status on this? I would love to use these in stable.

**Comment by xkr47** [comment]
Assume you have a function `getFoo()` returning `Option<String>` then it might be misleading if you do getFoo().contains("bar") since it looks like it checks for a substring when it in fact checks for an exact match. Would some other name perhaps be more clear?

**Comment by sinistersnare** [comment]
how about `contains_in` or `map_contains`?

**Comment by soc** [comment]
`option.contains_in("foo")` reads weird. You could of course pick something else entirely different, like `option.has("foo")` or `option.includes("foo")`.

But in the end, Rust is a typed language, so we should not be scared that much about confusion, because the types tell us which `contains` method is used.

**Comment by camsteffen** [comment]
Naming bikeshed. Sorry if this is the wrong place.

How about the names `eq_some`, `eq_ok` and `eq_err`? Or maybe only using those names for `Result`? I get that `contains` is consistent with containers, but that analogy weakens especially with `Result`.

**Comment by selendym** [comment]
I'm quite surprised bikeshedders have not suggested `wraps` yet, so I guess I'll be that person. ;)
This would be consistent with both `Option` and `Result`, and also `unwrap`.

**Comment by soc** [comment]
I think `wraps` has nice symmetry (though I wonder what's a good approach for `Result`: `wraps_err`/`err_wraps`?).

I haven't chosen it, because I think `unwrap` is a terrible name and probably will be deprecated in the long-term, so I don't want to add methods with names that could be an obstacle to that.

**Comment by frewsxcv** [comment]
if `unwrap` is going to be renamed in the long-term, why not rename a theoretical `Option#wraps` at the same time?

**Comment by genusistimelord** [comment]
you could write it as is(). 

then it would read as
getFoo().is("bar")

**Comment by xkr47** [comment]
or `has()`? :)

**Comment by genusistimelord** [comment]
> 
> 
> or `has()`? :)

Yeah Has would have the same meaning if it was a String comparison as you mentioned before. Though contains works great for this since we are looking for what is in the Some() and not looking at what is contained within the Variable contained within the Some() =3.

**Comment by jakoschiko** [comment]
Or
```
getFoo().is_some_and("bar")
tryFoo().is_ok_and("bar")
tryFoo().is_err_and(FooErr)
```
- The names are similar to `is_some`, `is_ok`, and `is_err`
- There are already functions that end with `_or`, e.g. `map_or`

**Comment by Rustinante** [comment]
Why is this feature even necessary? A lot of programming mistakes would be made due to the potential type confusion caused by this. Sometimes people will forget they're dealing with an `Option<HashSet<...>>` and call the `.contains` method on the option, thinking that it's actually a `HashSet`.

**Comment by Rustinante** [comment]
If some feature of this kind is really needed, I suggest calling it something like `is_some_eq`, where the `is_some_` prefix signifies that first of all it must be `is_some()` and then unwraps and compares under the hood.

**Comment by Rustinante** [comment]
@Centril please consider renaming this, the name collision will likely cause a lot of confusions and bugs down the road.

**Comment by jplatte** [comment]
I wouldn't mind it keeping the current name but since many are opposed to that, here's another idea: `.contains_some()`.

**Comment by Rustinante** [comment]
The word some has to come before contains, because first of all it must be a some, then a comparison is made, which is why at the minimum we should have something like some_eq. If we call it contains_some, it can be interpreted as saying the optional contains an optional value that is Some.

**Comment by camsteffen** [comment]
`x.eq_some(y)` reads like `x == Some(y)`

**Comment by ximon18** [comment]
I don’t mind `x.contains(y)` as conceptually I think it matches the goal well, and the clash with HashMap is (to me at least) not different from the fact that any type can have any fn and thus you have to know the type being operated on to understand the effect of the fn, but to differentiate it further from sounding like a collection test one could perhaps use:

 `x.contained_is(y)` and `x.contained_err_is(y)`

or:

  `x.inside_is(y)` and `x.inside_err_is(y)`

?

**Comment by Rustinante** [comment]
@ximon18 do you think `eq_some` is not good?

**Comment by ximon18** [comment]
@Rustinante: my concern with `eq_some()` is that while it might sound right for the `Option` use case, to me it sounds too much like it deals with an `Option` to be right for the `Result` use case.

**Comment by Rustinante** [comment]
@ximon18 for the `Result` there's the `eq_err` suggestion above.

**Comment by ximon18** [comment]
@Rustinante: True, but both `eq_some` and `eq_err` read (to me) like the value inside is `Some(y)` or `Err(y)` rather than actually being just `y`. Additionally by using `some` and `err` you duplicate something about the type, `Option` or `Result`, in the fn name which doesn’t feel right to me.

**Comment by Rustinante** [comment]
That objection would also apply to the existing function `is_some()` and `is_err()`

**Comment by ximon18** [comment]
@Rustinante: I see your point, and I’m not convinced that I’m right, but to me `is_some` and `is_err` don’t have the concern I mentioned because they don’t seem to be talking about the inside value while `eq_some(y)` and `eq_err(y)` are trying to say something about the inside value.

At any rate, I was trying to offer another possibility for consideration, not to comment directly on the value or otherwise of `eq_some` and `eq_err` as I’m not sure what is best or even if the proposal is actually necessary, especially with the new `matches!` syntax covering this use case (I think, if somewhat more verbosely).

**Comment by frewsxcv** [comment]
My take on `eq` is that it's misleading because we're using `PartialEq` in this method, not `Eq`

**Comment by Rustinante** [comment]
Then why do people still use the equality operator `==`  for both `Eq` and `PartialEq` comparisons? Isn't that misleading as well?

**Comment by Rustinante** [comment]
Then why do people still use `==` for both `Eq` and `PartialEq` comparisons? Isn't that misleading as well?

**Comment by frewsxcv** [comment]
> Then why do people still use `==` for both `Eq` and `PartialEq` comparisons? Isn't that misleading as well? And why can we still do `x == Some(y)`? I'm under the impression that this whole feature is to provide a shorter way for writing `x == Some(y)`.

This was covered earlier: https://github.com/rust-lang/rust/issues/62358#issuecomment-508900472

**Comment by Rustinante** [comment]
I see, but it still doesn't mean we can't use `eq`, even the trait method for `PartialEq` is `eq` https://doc.rust-lang.org/std/cmp/trait.PartialEq.html#tymethod.eq
Other proposed names don't even begin to touch on these nuances.

**Comment by frewsxcv** [comment]
Oh, nevermind then. `PartialOrd` has `partial_cmp`, but `PartialEq` has `eq`. Did not know that.

**Comment by soc** [comment]
@frewsxcv @Rustinante I wouldn't worry too much regarding `eq` & friends, the `Eq`/`PartialEq`/`Ord`/`PartialOrd` hierarchy makes little sense anyway.

**Comment by tvercruyssen** [comment]
What are still the blocking changes keeping this from being stabilized besides possibly a name change?

**Comment by soc** [comment]
> These methods are way too high up in the docs on both the `Option` and `Result` rustdoc. Would someone be able to send a PR to move these a lot further down? For example stuff like `ok` and `map` and `and_then` should all be featured higher than contains.

@tvercruyssen this perhaps?

**Comment by tvercruyssen** [comment]
> > These methods are way too high up in the docs on both the `Option` and `Result` rustdoc. Would someone be able to send a PR to move these a lot further down? For example stuff like `ok` and `map` and `and_then` should all be featured higher than contains.
> 
> @tvercruyssen this perhaps? shrug

If that's what is best, than yea I would be happy to make a PR. But aren't all methods ordered in alphabetical order? This would break that and might confuse people more (when scrolling through the docs looking for a method)?

**Comment by jhpratt** [comment]
The ordering of methods in documentation should have nothing to do with stabilization, as that can be changed at any time.

**Comment by tvercruyssen** [comment]
> I'm quite surprised bikeshedders have not suggested `wraps` yet, so I guess I'll be that person. ;)
> This would be consistent with both `Option` and `Result`, and also `unwrap`.

If we wanted to use this name instead that would have to happen before stabilization. That would move the method lower in the docs as well.

**Comment by camsteffen** [comment]
I would prefer that the "wraps" concept is ***not*** extended beyond `unwrap`. The `unwrap` function has a "do it the dangerous way" connotation, and so `wraps` is not really a correlary to `unwrap`. In fact, it could give a false impression that `unwrap` is more idiomatic than it actually is.

I still prefer `eq_some` over `contains`. One reason is that `get_items().contains(x)` is ambigious as to whether you are checking a collection type or an `Option`. The only argument against `eq_some` is that it is redundant with the type. But that is not true - the method checks that 1) it is the `Some` variant and 2) that the inner value is equal to the input. `eq_some` communicates this two-in-one semantic. In my mind, the semantics of this function and `Collection::contains` are "kind of similar but different", and its better not to conflate them.

**Comment by Rustinante** [comment]
+1 on @camsteffen's summary. I think we should do `eq_some` for `Option` and `eq_err` for `Result` based on the foregoing discussions.

**Comment by tvercruyssen** [comment]
So do we make a PR to rename `Option::contains` to `Option::eq_some`, `Result::contains` to `Result::eq_ok` and  `Result::contains_err` to `Result::eq_err` ? So that both methods can possibly(?) be stabilized at the same time.

**Comment by selendym** [comment]
> I would prefer that the "wraps" concept is not extended beyond unwrap. The unwrap function has a "do it the dangerous way" connotation, and so wraps is not really a correlary to unwrap. In fact, it could give a false impression that unwrap is more idiomatic than it actually is.

By that you are implicitly assuming that people are unable to read or comprehend documentation - if that were the case, all hope is already lost and this would be the least of our concerns. :)

Just to defend my bikeshed a bit:
- `unwrap` describes an _action_ that might fail with panic if it is not appropriate
- `wraps` describes a _proposition_ that is tested and simply returns the verdict

So there is a clear difference there, no need to underline (un)safeness any more than what the docs would have.
If anything, the compiler or linter could flag `unwrap` as evil, but that's another topic.

`wraps` is also easier to read, write, and remember than some long or multipart name; for example, consider:
```rust
if foo.wraps( bar ) {
    ...
}
```
This reads quite nicely, no?

In the end, this is just another round of bikeshedding.
As long as the name and meaning are easily remembered, anything reasonably sane should do.

**Comment by tvercruyssen** [comment]
Any sane name will do problem is more like how do we pick the right one. As a name change is seemingly(?) the only thing blocking an otherwise usefull feature from getting stabilized.

**Comment by jplatte** [comment]
I think `.wraps` might be a nice name for `(Option<T>, impl FnOnce(T) -> bool) -> bool`, i.e. the generalized form of `.eq_some` (assuming that `eq_some`, `eq_ok`, `eq_err` goes through).

**Comment by soc** [comment]
@camsteffen 

> One reason is that get_items().contains(x) is ambigious as to whether you are checking a collection type or an Option.

In what sense is `Option` anything but a collection with either 0 or 1 elements?

Also, the above code is also ambiguous whether it's a vector, a string, a slice, or a set. If the ambiguity was been an issue (it isn't) people would have hit it 15 years ago already (they haven't).

I propose consistency: come up with replacements for the other types too, so that things are properly disambiguated all the way! I propose `String::includes`, `slice::holds`, `Option::has` and `Vector::contains` in that regard.

(Or, simply use `contains`.)

**Comment by camsteffen** [comment]
"equals" is just a simpler concept than "contains" IMO.
1. "this contains a value equal to `x`"
2. "this equals `Some(x)`"

**Comment by soc** [comment]
@camsteffen So it would be `eq_slice`, `eq_string`, `eq_vector` too? I don't think the `Eq` type class is not flexible enough to accommodate that.

**Comment by camsteffen** [comment]
> @camsteffen So it would be `eq_slice`, `eq_string`, `eq_vector` too?

No?

**Comment by soc** [comment]
Then why the special-casing for `Option`?

The original commit literally had the message "This increases consistency with other common data structures." ... the naming is not a coincidence, it's the point.

**Comment by Rustinante** [comment]
@soc if the `Option` is for an underlying container class, it will easily lead to bugs because the one might accidentally confuse the `.contains` for the `Option` and the `.contains` for the underlying container. If you have two kids, why give them the same name?

**Comment by Rustinante** [comment]
`Option` is also different from the container classes. It participates in pattern matching in a special way in Rust and denote the **presence** or **absence** of values, while container classes don't. Similarly, the `Result` class has a special place in Rust, it denotes the **success/failure**. I hope we do not conflate them with the container classes.

**Comment by camsteffen** [comment]
Why do we have `is_none` instead of `is_empty`? That would be more consistent. Because `Option` is special. `eq_some`, `eq_ok` and `eq_err` are consistent with each other so that's worth something.

**Comment by tvercruyssen** [comment]
So we make a PR to rename all the `Option` and `Result` methods to `eq_some`, `eq_ok` and `eq_err` and reorder them maybe too ? Or are we still bikesheding names?

**Comment by camsteffen** [comment]
I think a lang team decision is needed at this point, if arguments have been exhausted.

**Comment by soc** [comment]
@Rustinante Is this a conjecture or is there any evidence for that? That is, did any of the languages that went with `contains` regret it later? Or is it a complete non-problem (which is the information I have from F#, Scala, ...).

**Comment by aaron-cerebras** [comment]
What’s a conjecture?

**Comment by Rustinante** [comment]
What’s a conjecture? And I’m not sure I get the point of comparing a specific function name in a different language, does Scala have a function called is_none too?

**Comment by Rustinante** [comment]
Also, speaking of regrets, yes i’m already regretting that we picked the contains name because of the reasons that were mentioned earlier in this thread, the confusion caused by the conflation of Option and containers was what led me to this GitHub issue in the first place. So I think there’s no need to look for further evidence of regrets in other languages…

**Comment by glittershark** [comment]
to put a little weight on one side of the discussion (though it might not be needed now) I actually *went looking* for `Option::contains` in my *autocomplete*, and was really happy to find it there, but disappointed to see that it was unstable. I'm pretty sure I'd never go looking for a function called `eq_some` - or any of the other names that have been proposed here. I know I'm just one person, but just wanted to provide a "field report" data-point of sorts in favor of the principle of least surprise.

**Comment by glittershark** [comment]
Also, rust already considers Option a container (as well it should) - look no further than `Option::iter()` and `Option::into_iter()`

**Comment by Rustinante** [comment]
Being able to iterate and being a "container" in the real sense are not the same thing. We can have a "container" that contains all the rational numbers but there won't be a natural way to iterate through it. On the other hand, being able to iterate through a generator of all the positive integers doesn't make the generator a "container".

It's true that nothing prevents us from making the Option a "container", but Option is a special language construct, as it participates pattern matching and is therefore more like an enum. Having different names such as `eq_some` would serve to remind us that we're dealing with an optional value, because, as I've mentioned before, it's very easy to confuse calling the `contains` on the underlying object in case it's a container, and calling the `contains` on the Option itself. And my field report would be that such confusions can easily lead to bugs.

**Comment by jhpratt** [comment]
@Rustinante I'm not sure if you're aware given your wording, but [`Option` _is_ an `enum`](https://doc.rust-lang.org/stable/std/option/enum.Option.html). There is nothing special about it with regard to pattern matching. `Option` isn't even a lang item, it just has some special casing for diagnostic purposes.

**Comment by aaron-cerebras** [comment]
I’m talking more in the context of functional programming, Option has a long tradition in those functional languages

**Comment by Rustinante** [comment]
I’m talking more in the context of functional programming, Option has a long tradition in those functional languages

**Comment by Rustinante** [comment]
And borrowing the unequivocal statement from @jhpratt , if we so unequivocally state that Options IS an enum, then the question is settled that Option shouldn’t be treated as a container, since making something both an enum and a container would probably just be confusing

**Comment by robjtede** [comment]
+1 on @glittershark's annecdote. I too have typed `.contains` and found it in the auto-complete dropdown, knowing immediately that it will do what I want (followed swiftly by disappointment that it is unstable).

It should probably be said that I did not stop to consider the nature of the Option container and it's facets as a psudeo-container type. It is simply a natural name for a method that checks some inner equality.

**Comment by ShadowJonathan** [comment]
I'll give some 2c regarding mental acrobatics as to why `::contains` is sensible here; `Option::Some` is a container, it "contains" the value `T` in itself, even though the container can also not contain it's value, with `contains` then you ask the question "`contains` this container value `X`?"

Some more 2c: even if this makes more sense to be `eq` after all, like @glittershark said, the muscle memory of `contains` is already there, so I'd argue this is more "missing" than anything else.

Lastly, I would personally find `thing.contains(x)` to be more readable than `thing == Some(x)` in long if-chains.

**Comment by Rustinante** [comment]
@ShadowJonathan im not sure if you realized but you just said Option::Some is a container, not Option itself

**Comment by soc** [comment]
@Rustinante I think you are really grasping for straws here, argument-wise.

Checking your argument against another language, e. g. Scala, none of your points feel like they are related to the naming of the method. If I inverted all those points ...

- `Option` is a library type with _some_ amount of special compiler treatment
- it could be an enum, but currently isn't
- you can pattern match on it, like on `Vector`, but it's considered poor style
- emptyness is being checked with `.isEmpty`, fullness with `isDefined`

... it wouldn't have any impact on the correct method name from my point of view.

Also, no one has reported being confused with `contains` there ever.

**Comment by Rustinante** [comment]
@soc I don't understand what you're trying to say. Nobody said anything about special "compiler" treatment. We have listed good reasons and not grasping for straws. Anybody interested please read the entire thread. There's no point arguing in a circle here.

**Comment by soc** [comment]
>> Nobody said anything about special "compiler" treatment

.

> but Option is a special language construct

.

> it just has some special casing for diagnostic purposes

**Comment by Rustinante** [comment]
By Option being a special language construct, i'm talking about Some/None in the functional programming sense, not whether it's a library or compiler treatment. The point being, as I have already said before, that Option is _different_ from other containers. If you like to look at other languages so much, see the Some and None in Ocaml and you'll see what i mean by Option being a different from vectors from a conceptual point of view.

For those that assert that it won't ever confusions, please do us all a favor and read the thread from the beginning, https://github.com/rust-lang/rust/issues/62358#issuecomment-676127715  etc.

And please let's have better reasons than citing "autocomplete"

**Comment by soc** [comment]
I think you just have to accept that 20 years of combined evidence of it being a non-problem is vastly more convincing than your imagination that it is.

And yes, being consistent, working with user expectations and playing well with auto-complete are all perfectly valid reasons.
Picking a name is largely about optimizing for readers, writers and completers, usually in that order.

**Comment by Rustinante** [comment]
20 years of combined evidence of what? In a different language there's a different context. We have listed several examples where confusions had shown up, those are already examples that show that it's not a non-problem.

**Comment by enterprisey** [comment]
> Assume you have a function `getFoo()` returning `Option<String>` then it might be misleading if you do `getFoo().contains("bar")` since it looks like it checks for a substring when it in fact checks for an exact match. Would some other name perhaps be more clear?

I was convinced by this argument, then I realized it also works for `Vec<String>`. (Minus some type inference shenanigans, see #42671.)

**Comment by kaleidawave** [comment]
Don't think this is mentioned in this thread but a "contains" method would be useful in situations in scenarios for result where `E` doesn't implement `PartialEq`. At the moment `my_result == Ok(42)` only works if `my_result::E: PartialEq`. I use a current `my_result.ok() == Some(42)` workaround occasionally (and there is the more verbose `if let Ok(v) = my_result { v == 42 } else { false }`). I think that `my_result.contains(42)` (or whatever the method name will be) has benefits over using the equality operator in this scenario.

**Comment by soc** [comment]
Looks stabilization is likely going to miss Rust 2021 edition.

**Comment by jhpratt** [comment]
@soc This change has nothing to do with the edition in the first place.

**Comment by m-ou-se** [comment]
This method might be more useful if it took a predicate:

```rust
option.contains(|c| c.is_lowercase())
```

For most cases, `== Some(x)` and `== Ok(x)` suffices can be used for checking (partial) equality, but checking whether an Option contains something fulfilling some predicate gets verbose or non-trivial:

```rust
let now: Time = ..;
let deadline: Option<Time> = ..;

// Different ways to check if the `deadline` has passed (if there's a deadline), all a bit verbose or confusing:
let a = deadline.map(|d| d > now).unwrap_or(false);
let b = deadline.map_or(false, |d| d > now);
let c = matches!(deadline, Some(d) if d > now);
let d = deadline.map(|d| d > now) == Some(true);
let e = deadline.filter(|d| d > now).is_some();
let f = deadline.iter().any(|d| d > now);
```

With a predicate-based `.contains`:
```rust
let g = deadline.contains(|d| d > now);
```

It wouldn't match the `.contains()` function we already have on slices and collections, but it would match `str`'s `.contains()` which can also take a predicate. I don't think consistency with slices is very important though. It's nice that `Option` and `Result` can be used with `.into_iter()` as a collection of 0 or 1 elements, but I don't think they need to match the full slice/collection interface.

**Comment by glittershark** [comment]
What if that was called `.any`, to match `Iterator::any`?

**Comment by soc** [comment]
@m-ou-se I think `str`'s design of `contains(Pattern)` is rather silly¹ and I don't think it makes for a good argument for picking one `Pattern` overload over the other – it wouldn't make it "match str's .contains() which can also take a predicate". Changing the API to take a `Pattern` would make it match `str`.

> I don't think consistency with slices is very important though.

I think it is. Having dealt with languages/libraries that were consistent and and with those that weren't, the difference is remarkable. Making users needlessly learn multiple APIs is not a good thing, it only adds to the mountain of things "you just have to know" which is way too high in Rust already.

Rust already suffering from inconsistent naming is not a carte blanche for introducing even more "naming creativity".

---

¹ It adds an additional mental indirection for the simplest case; I have never consciously seen it used in the wild; if I wanted to check against a predicate I wouldn't have thought of `contains`; looking at the `Pattern` API doc I wouldn't be able to use it.

**Comment by eddyb** [comment]
> What if that was called `.any`, to match `Iterator::any`?

The problem is that is that `.any`/`.all` imply generality over "arbitrary numbers of elements" in a way other methods `Option` and `Iterator` share (`.map`, `.filter`, etc.) generally don't.

Closest equivalents I can think of would be `.is` or `.has`, as singular counterparts for "`all` are" / "`all` have".

**Comment by camsteffen** [comment]
I don't have any qualms with `str::contains(Pattern)`, but I do fear it might be problematic to use it to justify `Option::contains` with a predicate. In my mind, the name `str::contains` makes sense primarily because it accepts `&str` or `char`, and the char predicate is an added convenience that avoids adding another method. This would be the first `contains` method that *does not* accept a value to compare against and I think that would be perplexing (though not terribly so).

That said, I agree that the predicate version seems more generally useful. So I think it would be good to pursue that with higher precedence than the PartialEq version.

Some ideas: `some_and`, `and_is`, `is`, `where`, `having`, `and_thusly`<-joke

> For most cases, `== Some(x)` and `== Ok(x)` suffices can be used for checking (partial) equality,

So does the predicate use case even overlap with the PartialEq use case? It seems like there is theoretically room for two functions here. But again, I'd pursue the predicate function first and see where that leaves us.

**Comment by ShadowJonathan** [comment]
I personally like `and_is` the most, and for the predicate option, i'd like to suggest `and_with`.

(Though something in the back of my mind tells me `_with` was for with a return value generic, not `bool` returns)

**Comment by m-ou-se** [comment]
> It seems like there is theoretically room for two functions here.

Maybe, but each function will need a clear name. And I'd really like to use `contains` for `.contains(|x| x > 5)` rather than `.contains(&5)`, since I think that the latter will rarely be useful (especially for `Option`s).

We tend to have a relatively high bar for additions to fundamental things like `Option` and `Iterator`. I personally think `.contains(&5)` does not pass that bar, but `.contains(|x| x > 5)` does.

(Side note: I also think the existing `.contains()` for slices was a mistake, and should've taken a predicate too. But at least there's `.iter().any(..)`, which for slices is quite obvious and readable. (Unlike for `Option`s and `Result`s.))

**Comment by m-ou-se** [comment]
> I personally like `and_is` or `some_is` the most, and for the predicate option, i'd like to suggest `and_with`/`some_with`.

If I look at these three lines of code
```rust
let a = optional_char.contains(|c| c.is_lowercase());
let b = optional_char.and_with(|c| c.is_lowercase());
let c = optional_char.some_with(|c| c.is_lowercase());
```
then as an reader without context I'd probably only understand the first one without having to look up the documentation to see what the method does.

`and_with` makes me think of `and_then`, which is quite different. And `some_with` reminds me of `bool::then_some`, and sounds like it'll produce a `Some` rather than a boolean.

**Comment by ShadowJonathan** [comment]
Good points, though still, when i personally look at `.contains()` i more or less expect a `X contains Y` "english" wording, rather than `X contains (test with Y)`.

An alternative (that i forgot to write down) was `and_test`, or just `test`, but that might touch the entire testing idea of rust, so i wasn't 100%

(Also sorry if this bikeshedding is not warranted at the moment)

**Comment by jhpratt** [comment]
In my opinion anything that accepts a closure here should have "if" in the name.

**Comment by camsteffen** [comment]
`and_if` complements `and_then` nicely.

**Comment by soc** [comment]
Perhaps we can split the bike-shedding of unrelated topics into its own issue and consider what we can do to get this one across the finish line?

**Comment by ShadowJonathan** [comment]
I dont think that any bikeshedding of unrelated topics have come up, this is still primarily about the naming of `contains` specifically, afaik.

**Comment by gorsat** [comment]
Such a useful and intuitive feature with such a straightforward implementation yet it's been stuck here for 2 years? Wow... 

As someone who is new to Rust I expected this feature to exist for Option, and I happily and quickly found it in the docs, but then discovered I'd have to switch to unstable to use it. Now I come here and find it has been stuck in a naming debate for literally 2+ years. That's really unfortunate. Doesn't the Rust community have a way to push through this log jam?

**Comment by univerz** [comment]
i'm happily using nightly feature, `contains` + `cointains_with` for lambda case looks intuitive & matches behaviour of other _with apis.

**Comment by emi2k01** [comment]
Someone mentioned `test` but they dismissed it in the same comment but I think it's a good one.

`opt.test(|n| n > 10)`
`res.test(|n| n == 20)`
`res.test_err(|e| e == ErrorKind::NotFound)`

The word "test" is already used to describe similar methods like `Iterator::any` and `Iterator::all` that take a closure and return a boolean. It is also clear what is doing.

I think any of the already mentioned names by other people are clear enough and most people should understand what the method does given the context on where it appears.

```rs
if file.metadata().method(Metadata::is_file) {
    std::fs::copy(&from, &to)
}


if string.parse_url().method(|url| url.has_host()) {
    urls.push(string);
}
```

Replace `method` with any of the suggested names above and any programmer would know what the code is doing.

I really doubt the name of these functions are worth discussing for more than two years.

**Comment by pmnoxx** [comment]
> Someone mentioned `test` but they dismissed it in the same comment but I think it's a good one.
> 
> `opt.test(|n| n > 10)` `res.test(|n| n == 20)` `res.test_err(|e| e == ErrorKind::NotFound)`
> 
> The word "test" is already used to describe similar methods like `Iterator::any` and `Iterator::all` that take a closure and return a boolean. It is also clear what is doing.
> 
> I think any of the already mentioned names by other people are clear enough and most people should understand what the method does given the context on where it appears.
> 
> ```rust
> if file.metadata().method(Metadata::is_file) {
>     std::fs::copy(&from, &to)
> }
> 
> 
> if string.parse_url().method(|url| url.has_host()) {
>     urls.push(string);
> }
> ```
> 
> Replace `method` with any of the suggested names above and any programmer would know what the code is doing.
> 
> I really doubt the name of these functions are worth discussing for more than two years.

How about using the word `any`? 

Right now, you can use the world `any` on iterator 
```
   [1,2,3].iter().any(|x| x == 2)
```

This also works on iterators from options:
```
   Some(2).iter().any(|x| x == 2)
```
Using the same name for option, directly, would be more discoverable.
```
   Some(2).any(|x| x == 2)
```

**Comment by ShadowJonathan** [comment]
> When I started thinking about it. Option<...> is just an iterator for at most one element.

It's been noted earlier in the thread that there are both stances of looking at `Option` as an iterator, and looking at it as it's own thing/`Result`-ish, just fyi.

**Comment by pmnoxx** [comment]
> > When I started thinking about it. Option<...> is just an iterator for at most one element.
> 
> It's been noted earlier in the thread that there are both stances of looking at `Option` as an iterator, and looking at it as it's own thing/`Result`-ish, just fyi.

In my opinion current apis for `Option` / `Result`, `Iter<Val>` / `Iter<Result<Val, Error>` are a bit inconsistent.
I found out that I have to learn separate naming for all of those types above. Methods even though they do the same thing ,they are called differently.

I've thought a bit, on how we could make an easy to remember api, that works for both options, results, iterators, iterators of results. Here is what I came up with `map`

```rust
      /// Vec<Result<...>>
        let mut vec: Vec<Result<u32, Error>> = Vec::new();

        // modifies Ok(...)value  // doesn't exist  yet
        let x = vec.iter().map_ok(|x| x).collect_vec();
        // modifies Err(...)value  // doesn't exist yet
        let x = vec.iter().map_err(|e| e).collect_vec();
        // this modifier the value itself // that's how it currently works
        let x = vec.iter().map(|v| v).collect_vec();

        /// Result
        let mut x: Result<u32, Error> = Ok(5);
       
        // modifies Ok(...)value  // doesn't exist  yet
        x.map_ok(|ok|ok) 
        // modifies Err(...)value  // exists, works as expected
        x.map_err(|e}e) 
        // Should modify  the result // exists, but modifies ok value
        x.map(|r||r) 

      /// Vector of values
        let mut vec: Vec<u32> = Vec::new();
      // map exists; works as expected
        let x = vec.iter().map(|v| v).collect_vec();

        let mut x: Option<u32> = Some(5);
        // maps value // exists; works as expected
       x.map(|x| x * 5)
```

My argument would be to have similar api for `any`, just make it work with `any`, `any_ok`, `any_err`, just as seen above.
I really, don't want to spend time learning 3 types of api names for `options` / `results`, `iter`, and also not having good `Iter<Result<...>>` options.

I found out white processing values in iterators that there is no good error handling.

I really would like to write something like this, for processing network queries.

```rust
fn url_query(id: usize) -> Result<UrlData, Error>()

(0...1000)
.map(|id|url_query(id)) // now current data isResult<UrlData, Error>()
.map_err(|e|pretty_print(e))
.map_ok(|v|process(v))
.collect()
```

Currently I would have to do the following
```rust
(0...1000)
.map(|id|url_query(id)) // now current data isResult<UrlData, Error>()
.map(|v|{
   if let Err(e) = &v{
       pretty_print(e))
   }
   v
}
.map(|v|{
   if let Some(val) = &v{
       Some(process(val))
   } else {
      v
   }
}
.collect()
```

**Comment by soc** [comment]
> Right now, you can use the word `any` on iterator

None of the methods in question take a lambda.

**Comment by mbartelsm** [comment]
I have to say I agree with @pmnoxx comment. Currently, the APIs for Option, Result, and Iter are all over the place. You can argue that they are not the same and should not be the same, but in practice, they all serve and are used as Monads whose inner values can be modified in consistent ways (albeit, with inconsistent APIs).

As a relatively new user, one of the most annoying parts of Rust is having to go back and forth between code and documentation for types whose behavior ought to be the same.

In that regard, I think using closures and naming this `any` would be the most consistent approach. It might impose a small barrier of entry for learning what `any` does since it might not be immediately apparent, but it has the overwhelming benefit that you only need to learn the method name once for three of the most common types/traits in the std lib.

**Comment by m-ou-se** [comment]
> How about using the word `any`?

I disagree with trying to make `Option` look like an iterator. We already have a way to view an Option as an iterator: `option.into_iter()`, which gives a `std::option::IntoIter`: an Option wrapper with an Iterator interface. While Option is isomorphic with an `option::IntoIter`, they're separate types on purpose. `something.any(f)` makes it look like `f` might be executed multiple times, which isn't the case for a contains/any/test method on `Option`. (Also note that slices/arrays etc. don't have an `.any()` method either.)

> Someone mentioned `test` but they dismissed it in the same comment but I think it's a good one.

The name `test` doesn't make it clear what happens in the `None` case. For example, in `optional_timeout.contains(|t| t > now())` reads to me as "Does it contain a timeout that has passed?", to which the answer is clearly no/false in the case of `None`. This is not as clear with a less meaningful name like `test`.

**Comment by soc** [comment]
@mbartelsm You can divide these types roughly into data structures (i. e. things that usually hold concrete values: strings, arrays, slices, vecs, maps, sets) and descriptions of computation (i. e. things that describe how values should be processed: Iter in Rust, Stream in Java, etc.). They are closely related, but not the same – for good reasons.

I added `contains` because `Option` and `Result` are closer to the first category (strings, arrays, slices, vecs, maps, sets) and all of those have `contains` with the same (or similar) signature.

Whether `any` should be added to `Option`/`Result` is not something I particularly care about, but I think `any` would neither be more consistent than `contains` nor closer to the intuition of users.

**Comment by m-ou-se** [comment]
@rust-lang/libs-api Any opinions on the discussion starting at https://github.com/rust-lang/rust/issues/62358#issuecomment-970280621 ?

**Comment by BurntSushi** [comment]
Briefly, for me, I'll say that I am not a fan of using `any` here. Largely for the same reasons [pointed out by @m-ou-se](https://github.com/rust-lang/rust/issues/62358#issuecomment-1005681278) earlier. It seems like a highly confusable name to me.

I also think whatever routine we settle on should probably accept a predicate. I think a routine that accepts a value based on `PartialEq` isn't nearly as useful.

I'm not a _huge_ fan of the name `contains`, but I think it fits. It's also hard for me to think of a better name. I somewhat like `some_is`, `ok_is` and `err_is`, where `is` is in my experience strongly associated with "predicate." But the names don't feel quite right to me for reasons that I'm finding difficult to articulate.

I think for me, with respect to `contains`, I do find the _near_ consistency we have with that API in std to be appealing. It is true that `str::contains` is polymorphic and can accept a predicate, but I _think_ every other use of `contains` accepts a value of some sort, and not a predicate. That does seem like a nice consistency to preserve and codify if possible.

**Comment by mbartelsm** [comment]
While I did advocate for using `any` instead of `contains`, I do want to say that I find it more important to have this method with some name than to not have it, for the reasons that @m-ou-se points out in their comment. It's verbose to write and with the current stable methods and many of the ways to write it obscure legibility.

**Comment by ShadowJonathan** [comment]
One last name suggestion; `has_with`, where `has` alludes to the "contains", and `with` implies a predicate.

**Comment by Rustinante** [comment]
@BurntSushi what do you think about the alternatives that were brought up earlier, eq_some, eq_err?

**Comment by yaahc** [comment]
I mentioned this in the libs team meeting but I'll reiterate here, but I personally don't agree that `deadline.map(|d| d > now).unwrap_or(false)` looks verbose or confusing, and I prefer it over `deadline.contains(|d| d > now)`.

I'm trying to think through why this is, and as an exercise I thought I'd try imagine how I'd read these expressions if they were part of an if statement. In the process I realized I wouldn't read the former as sentence, and would instead want to read it as it's own boolean producing expression that I would then use in the if. So if I were to use these in practice I think I would write slightly different code.

```rust
// kinda had to guess at what this snippet was representing
let deadline_is_active = deadline.map(|d| d > now).unwrap_or(false);
if deadline_is_active {
    // ...
}
```

vs 

```rust
if deadline.contains(|d| d > now) {
    // ...
}
```

Admittedly, one could still outline and name the condition in the second example the same as the first, and the difference would be substantially reduced, but do think that the former snippet is more likely to encourage this style of separation. My conclusion is that you're absolutely right @m-ou-se that `contains` would be less verbose, but I'm still not at all convinced that it improves readability.

**Comment by avirule** [comment]
> 
> 
> ```rust
> if deadline.contains(|d| d > now) {
>     // ...
> }
> ```
> 
> This latter one on the other hand I do feel comfortable reading directly, and I'd read this as "if deadline contains some instant later than now".
> 
> Admittedly, one could still outline and name the condition in the second example the same as the first, and the difference would be substantially reduced, but I do think that the former snippet is more likely to encourage this style of separation. My conclusion is that you're absolutely right @m-ou-se that `contains` would be less verbose, but I'm still not at all convinced that it reduces confusion.

In my opinion, the latter is confusing—at least, in my opinion, `.contains()` is mostly relevant when working with slices or collections. I've been programming for several years, and I can honestly say I've never desired `.contains()` to take a closure rather than a direct equatable type. This has been consistent through the several languages I've become somewhat familiar with: C++, C#, and Rust.

Which isn't to say my experience means this is a proper assertion, but rather that my experience with these languages has somewhat imparted the idea that `.contains(other)` semantically equates to:
```rust
for obj in collection {
  if obj.eq(&obj_other) {
    return true;
  }
}
```
Effectively, a direct equality comparison.

So, to my mind, taking a closure in `.contains()` is a move that would distance Rust's Option<> API from what a common programmer would immediately assume of it. I do believe we should—where possible—strive maintain a consistency with the established conventions present in existing languages.

I, for one, am a fan of `some_eq` or `some_is` semantics. This convention would closely follow existing APIs (i.e. `map_ok`, `map_err`, `is_some`) in their deference to the unwrappable state of the type, while additionally following existing semantics regarding equality comparison naming conventions (which I hope I've expressed is what `.contains()` indicates is required).

**Comment by chris-zen** [comment]
I suggest looking at how other functional languages solve this same problem. In the case of Scala, the interfaces between `Option`, `Either` and other collections are quite consistent and they provide the following operation:

```scala
def exists(predicate: (B) => Boolean): Boolean
```

I personally read `maybe_value.exists(|value| value > 5)` as `exists a value such that it is greater than 5`.

Not willing to introduce yet another name, but to remark the importance for it to help the reader understand something like `exists a value such that <predicate>`.

I'd also like to stress that beyond the name that it is chosen, this feature will be really useful, and has been missing for a log time :wink:

**Comment by ShadowJonathan** [comment]
`exists` has not been suggested before in this thread, I'm heavily in favour of it, it alludes to a predicate much better than `contains` does. :+1:

**Comment by avirule** [comment]
> I suggest looking at how other functional languages solve this same problem. In the case of Scala, the interfaces between `Option`, `Either` and other collections are quite consistent and they provide the following operation:
> 
> ```scala
> def exists(predicate: (B) => Boolean): Boolean
> ```
> 
> I personally read `maybe_value.exists(|value| value > 5)` as `exists a value such that it is greater than 5`.
> 
> Not willing to introduce yet another name, but to remark the importance for it to help the reader understand something like `exists a value such that <predicate>`.

I believe, for the same reasons `.eq()` doesn't take a closure, it would be unwise to do so for this API. At its heart (at least, in the spirit of the functionality I've grasped), it is just a shorthand for a conditional equality comparison.

**Comment by soc** [comment]
As there seems to exist some severe confusion about the method in question, here is the signature again for your convenience:

```
pub fn contains<U>(&self, x: &U) -> bool where U: PartialEq<T>
```

**Comment by avirule** [comment]
> @avirule This tracking issue hast 69 upvotes.

It also has 120 comments, a large portion of which are related to bikeshedding the naming. Admittedly, I don't spend much time in issues or PRs, but it seems to me that the community agrees `.contains()` wouldn't align with the semantics of `Option<>` and `Result<,>`. I'm inclined to agree, given the aforementioned methods are generally used in the contexts of collections, which the two types definitely *are not*.

I believe, for example, it's widely seen as a mistake to have shared so many functions between `Option` and `Iterator`, as it leads to confusion (or entirely misrepresents) what happens to the type behind-the-scenes. And so, it's with that in mind that I've tried to move this issue's bikeshedding along, rather than suppress it—to me, there is a clear need to disassociate `Option<>` and `Result<,>` from collection types, such as `Iter`, `Vec`, slices, etc.

**Comment by soc** [comment]
> It also has 120 comments

The 69 upvotes each stand for a distinct person, and they all seem to generally approve of this PR.

The 120 comments on the other hand are produced by a minority that can't even agree with itself whether the problem with `contains` is that it is too consistent and should be made more inconsistent – or that it is too inconsistent and should be made more consistent. And that's not even taking the value-vs.-lambda bikeshedding into account.

> it's widely seen as a mistake to have shared so many functions between `Option` and `Iterator`

`Iterator` does not provide a `contains` method. And even if it was, I'm not sure on what "widely seen as a mistake" is based on.

> there is a clear need to disassociate Option<> and Result<,> from collection types, such as Iter, Vec, slices, etc.

I don't see that "clear need", and I can back that up with 10 years of Scala shipping with `Option::contains`. In that time, how many people came up to me and reported an issue with this design (not just some conjecture of what might happen)?

Not a single one.

**Comment by Rustinante** [comment]
I came up with reported issues very early on in this discussion if people care to go back and read

**Comment by glittershark** [comment]
>  [Option and Result] definitely _are not_ [collections]
Can you expand on why you think this is the case? Every criterion that I can imagine that qualifies a type as a collection applies to both Option and Result

**Comment by soc** [comment]
@Rustinante And I dismissed them, because making assumptions about what might happen is not equal to a decade of actual evidence to the contrary.

**Comment by Rustinante** [comment]
They were actual bugs, not assumptions

**Comment by ShadowJonathan** [comment]
> In that time, how many people came up to me and reported an issue with this design (not just some conjecture of what might happen)?
> 
> Not a single one.

That is not as decisive as a data point as you might think, language design is subtle, for anyone who might not be distinctly aware of it's intricacies, or any alternatives, they'd either dont know which grass would be greener, or would simply have a vague feeling of "wrong" of all the small decisions that accumulated in their experience.

Rust especially prides itself with being correct and "understandable", this is one such instance, i think that @avirule put it right with that "we feel this is wrong", i personally am just trying to find *an* alternative to `contains` for this discussion, as, while i think it works plenty well for a "`y` is a box, `x` is inside of it" scenario, it doesn't feel right with a predicate.

---

> And I dismissed them, because making assumptions about what might happen is not equal to a decade of actual evidence to the contrary.

Please keep in mind that, while your experience in Scala may be very good with applying Scala ideology to Scala problems, Rust might not fit in the same corner. Superficially, the types may look and work the same, but i think that "experience" here doesn't translate as distinctively and decisively as you make it out to be, while you may have an opinion, i don't think you have an authority in this decision, and especially not one to ignore experiences from other users.

**Comment by soc** [comment]
> They were actual bugs, not assumptions

@Rustinante Could you share the links to the (at least two) bug reports we should clearly all be aware of?

> it doesn't feel right with a predicate

`contains` does not take a predicate.

> while you may have an opinion, i don't think you have an absolute authority in this decision, and especially not one to ignore experiences from other users.

The "opinion" is backed up by facts of a standard library shipping this design to its users for 10 years.

I'd love to learn more about these "experiences", but at this point people had plenty of time to come up with _anything_ that would support it, but so far nothing has materialized in that regard. I have high hopes of @Rustinante coming through, though.

**Comment by ShadowJonathan** [comment]
> contains does not take a predicate.

A large part of the conversation has been to change `fn contains(&T)` to a `fn contains(f: impl FnOnce(&T) -> bool)`, i was talking about this part.

> The "opinion" is backed up by facts of a standard library shipping this design to its users for 10 years. 

In Scala, this is Rust.

> but I think it's fair to argue for the design the overwhelming majority seems to be perfectly fine with.

That is not what is happening here, the bikeshedding has shifted away from "contains with value" to "contains with predicate", this is not reflected in the initial post, as i believe *that* is what most people are upvoting.

**Comment by soc** [comment]
> In Scala, this is Rust.

If you have more applicable experience of actual usage of such a method in the wild, then please share it.

>  this is not reflected in the initial post, and so i believe that is what most people are upvoting

Why would people upvote a PR about `contains(<value>)` if they actually wanted `contains(<lambda>)`?

> I'm fine with contains(&T), im not fine with contains(fn(&T) -> bool).

Great! Then let's get this over with, and the people who want to have a method that takes a lambda can have their own PR and all the bike sheds they want in it.

**Comment by Rustinante** [comment]
As a user of Rust, why should I have to think about whether what I pass to `contains` should include the Option **and** the value inside the Option, or just the thing inside the option? Why don't we just use another name so that this mental gymnastics never has to be done??? This mental gymnastics is even worse with the predicate.

**Comment by soc** [comment]
> As a user of Rust, why should I have to think about whether what I pass to contains should include the Option and the value inside the Option, or just the thing inside the option?

As thousands of users of Scala have shown, you don't. The problem does not exist.

**Comment by Rustinante** [comment]
As thousands of users of Scala have shown, Rust does not need to exist.

**Comment by ShadowJonathan** [comment]
> Why would people upvote a PR about contains(<value>) if they actually wanted contains(<lambda>)?

Apologies, i meant that people upvote for value, not lambda, i had some confusing wording, i fixed it.

> Great! Then let's get this over with, and the people who want to have a method that takes a lambda can have their own PR and all the bike sheds they want in it.

Its not that easy, @m-ou-se [expressed an opinion](https://github.com/rust-lang/rust/issues/62358#issuecomment-970280621) to change it from value to lambda directly, so i don't think the current API will stabilise as-is.

> As thousands of users of Scala have shown, you don't.

Since when have Scala users started thinking like Rust users? Wit aside, i think that you are placing too much weight on your experience with Scala specifically to dictate other's opinions. I would like to see some examples of how Scala handles all of this, as i asked in a [previous comment](https://github.com/rust-lang/rust/issues/62358#issuecomment-1013938718).

**Comment by Rustinante** [comment]
If you just want to carbon copy the ideology of Scala into Rust, and are not interested in improving on the language, I suggest you go back and work on Scala

**Comment by soc** [comment]
@Rustinante Please calm down, this is not a good luck.

**Comment by avirule** [comment]
> > [Option and Result] definitely _are not_ [collections]
> 
> Can you expand on why you think this is the case? Every criterion that I can imagine that qualifies a type as a collection applies to both Option and Result

You *cannot* iterate `Option<>`s or `Result<,>`s, even if they can be coerced into an iterator (this is likely a convenience measure, rather than a purely design-focused choice). Additionally, the types cannot be indexed, as every collection is expected to be able to do (in some form or another).

At first glance, what's proposed here makes sense. `Option<>` can be thought of as a container, with either something or nothing in it. So, to say `option.contains(stuff_that_goes_in_a_container)` or `does this container contain this value?` makes sense. I'll outline, hopefully more clearly, the issues I see with this reasoning:

- `Option<>` isn't a container. It's a wrapper (`.unwrap()`..), around a single object. Where you can put as many items as you want (or can fit) into a container, an `Option<>` will only ever hold 1 or None objects.
- Similarly, given the association `.contains()` has with lists, this would further suggest, semantically, that `Option<>` could somehow hold multiple values.
- `Option<>` cannot be iterated, and this is where my issues with it sharing so many methods with `Iterator` comes from—relevant to this discussion, however, is the desire for `Option<>` to not stray even further towards being treated as a collection.
- The semantics of `Option<>` don't support this kind of operation, with this kind of naming. To actually use the value within an option, you're expected to somehow unwrap it, or acknowledge that it's `Some()` or `None`, i.e. `.is_some()`, `.expect()`, `.map()`, `let Some(_) = option`. These force you to implicitly understand the option is *wrapping* something, and what you're doing is *unwrapping* it.

> I don't see that "clear need", and I can back that up with 10 years of Scala shipping with `Option::contains`. In that time, how many people came up to me and reported an issue with this design (not just some conjecture of what might happen)?
> 
> Not a single one.

I don't know if you regret writing this or not, but this is an awful argument. Now, I've never used Scala—but Rust isn't Scala. One of the core tenets of Rust is to do things *right*, and *not* make the same mistakes other languages have made.

So, to finalize: Does this, in the grand scheme of things, add up to much? Certainly, no. One method won't make or break a language. But as we've all experienced with countless languages, it's these little inconsistencies that can add up to what makes a confusing, difficult-to-grok developer experience.

**Comment by soc** [comment]
> Wit aside, i think that you are placing too much weight on your experience with Scala specifically to dictate other's opinions. 

@ShadowJonathan I hope you see the utter hypocrisy in that?

I'm offering evidence to have a fact-based discussion and get criticized for it, but at the same time "opinions" need to be treated as gospel? Sorry, opinions and facts don't have the same merit. If you don't like the evidence I provide, please bring your evidence own to the table instead of complaining that you don't like mine.

> I don't know if you regret writing this or not, but this is an awful argument. Now, I've never used Scala—but Rust isn't Scala. 

@avirule It's not an awful argument, it's a very good one. It is based on learning from other languages. Pretending that Scala and Rust have nothing in common and that it's impossible to apply the lessons learned from the former to the latter is pretty anti-intellectual in my opinion. Like it or not, Rust's Option took more inspiration from Scala's Option than from any other language (including OCaml's Option), which is why comparing it with the former makes a lot of sense.

> One of the core tenets of Rust is to do things right, and not make the same mistakes other languages have made.

I could probably write a book about things-wrong-in-Scala, but having consistent naming is certainly not one of its chapters.

> But as we've all experienced with countless languages, it's these little inconsistencies that can add up to what makes a confusing, difficult-to-grok developer experience.

`contains` as proposed is consistent with Rust's existing naming and design choices.

**Comment by Rustinante** [comment]
Evidence can be either empirical, or a priori, Analytical reasoning is not the same as opinions.
I want to point out one thing about the “container” for those who still have common sense left in them. It is the “Some” entity that can be said to “contain” something (even then wraps would be a better word), not the Option entity. Calling the “contains” on the Option entity would be to skip one level of abstraction, and ignoring the None variant, which is extremely confusing.

**Comment by soc** [comment]
> for those who still have common sense left in them

@Rustinante not sure writing things like this is helping your argument.

What would certainly help your point of view though – is providing links to the actual bugs you encountered.

**Comment by jakoschiko** [comment]
I've learned Scala for my current job. At the beginning I was surprised that `Option` was treated as a container, but I got used to it very quickly. My conclusion after using it for several years now: It's convenient when writing code, but I would agree that it hurts readability. For example, I prefer this
```rust
// Rust
if let Some(result) = calculate() {
  ... // This code will be executed at most once
}
```
over this
```Scala
// Scala
calculate().foreach { result =>
  ... // How often will this code be executed?
}
```
Treating `Option` as a container has pros and cons. If Rust has a focus on readability, it's totally valid to avoid function names used by container types.

**Comment by avirule** [comment]
> > I don't know if you regret writing this or not, but this is an awful argument. Now, I've never used Scala—but Rust isn't Scala.
> 
> @avirule It's not an awful argument, it's a very good one.
> 
> It is based on learning from other languages. Pretending that Scala and Rust have nothing in common and that it's impossible to apply the lessons learned from the former to the latter is pretty anti-intellectual in my opinion.
> 
> Like it or not, Rust's Option took more inspiration from Scala's Option than from any other language (including OCaml's Option), which is why comparing it with the former makes a lot of sense.
> 
> > One of the core tenets of Rust is to do things right, and not make the same mistakes other languages have made.
> 
> I could probably write a book about things-wrong-in-Scala, but having consistent naming is certainly not one of its chapters.
> 
> > But as we've all experienced with countless languages, it's these little inconsistencies that can add up to what makes a confusing, difficult-to-grok developer experience.
> 
> `contains` as proposed is consistent with Rust's existing naming and design choices.

I've tried to outline my reasoning on this topic to the best of my ability, so at this point I believe anything more I have to say will just be reiteration. I'm still of the mind you've failed to address my reasoning with any substantial arguments, but I don't control this issue, and if your mind can't be changed—well, there's not much to be done about it. I support the essence of this issue (of what it seeks to achieve), even if I don't support the semantics.

For what it's worth, I wouldn't suggest throwing around your experience working on Scala as much as you do. Our pasts can inform our present and future, but they do not define them—if your ideas are to have merit, they must be able to stand on more than themselves. They will appeal to logic and reason, rather than to your own authority of the subject.

**Comment by m-ou-se** [comment]
This was discussed recently in the library team meeting. We'll investigate what it takes to have `option == Some(x)` working for `option: Option<T>` and `x: U` where `T: PartialEq<U>`, which would remove the need for a `PartialEq`-based `contains` function. (Same for `result == Ok(x)`.) If we can make that work as intended, possibly with a small language change, that is preferred over a method for doing the same thing.

A version taking a predicate seems generally useful, but it is unclear if `contains` is the best name for it. Since this combined discussion is starting to get a bit out of hand, I'll send a PR to add the predicate version with (for the time being) the name `is_some_with` (and `is_ok_with`), with its own tracking issue.

**Comment by khionu** [comment]
I'm saddened by how this thread turned towards attacking character, but even more so that it took so long for anyone to break the cycle. @soc @Rustinante, you both acted in pretty poor faith. It takes two to escalate, but only one to deescalate.

@Rustinante we don't tell people they should leave the community, ever. If there's a legitimate reason for someone to leave the community, that has to go through the Mods Team.

@soc What avirule said about your experience with Scala is correct. While potentially very valuable to reflect on while contributing, it isn't merit for the points in of itself. This is why we have a Prior Art section in the RFC template: it's great to have references, but they are only supportive of already solid reasoning.

Please consider this a formal warning, and please try to contribute in good faith from now on.

**Comment by dustypaws** [comment]
Now......can we stabilize this feature? Maybe...? 🥺👉👈

**Comment by soc** [comment]
I'm unsubscribed, as this is a poor use of my time.

I ask everyone to kindly stop constantly pulling me back into this thread (for instance by @-mentioning me), please.

**Comment by krdln** [comment]
Note on `contains` naming – this is affecting compiler messages even though the feature is not even yet stabilized! See https://github.com/rust-lang/rust/issues/95428 – I've recently encountered someone getting really confused on an error (which turned out to be missing `unwrap()`).

**Comment by DaniloArruda** [comment]
I believe that just "is", instead "contains" could be a better name for this funcionality

**Comment by camden-smallwood** [comment]
Please merge this, I'm tired of using `map_or` and explicitly supplying `false` and a lambda just to check for equality on the contained value...

**Comment by arniu** [comment]
How about `map_or_default`?

**Comment by camden-smallwood** [comment]
I can't find an `Option::map_or_default`, but I would assume from the implication of the name that it would make it so that I no longer have to explicitly supply `false`. The lambda is still there however, making the logic still appear indirect and not exactly solving the problem at hand:

I have an `Option<T>` stored in a variable `x`. I want to check if this `x` has a value of `Some(T)` and compare the inner `T` to some other `T` stored in a variable `y`. 

If `Option::contains` were stable, then I could do the following, which falls more in line with the intent and is more preferable in my opinion:

```rust
if x.contains(&y) {
    // ...
}
```

But since `Option::contains` is not stable, I have to choose from a set of workarounds which are less clear and less ideal. For example:

```rust
if let Some(x) = x {
    if x == y {
        // ...
    }
}
```

```rust
match x {
    x if x == y => {
        // ...
    }
    _ => {}
}
```

```rust
if x.map(|x| x == y).unwrap_or(false) {
    // ...
}
```

```rust
if x.map_or(false, |x| x == y) {
    // ...
}
```

**Comment by arniu** [comment]
@camden-smallwood `if x == Some(y) { ... }` works

**Comment by Buckram123** [comment]
> > @camden-smallwood `if x == Some(y) {}` works, and `if let Some(x) = x && x == y {}` works too.
> 
> The first one does indeed work on stable, thanks. The second one isn't stable unfortunately...

but `y` gets moved if not `Copy`, so it's not same

**Comment by bjorn** [comment]
What about `has_value` instead?

**Comment by arniu** [comment]
```rust
impl<T: PartialEq> PartialEq<T> for Option<T> {
    fn eq(&self, other: &T) -> bool {
        match self {
            Some(this) => this == other,
            None => false,
        }
    }
}
```

**Comment by FHTMitchell** [comment]
> > > @camden-smallwood `if x == Some(y) {}` works, and `if let Some(x) = x && x == y {}` works too.
> > 
> > 
> > The first one does indeed work on stable, thanks. The second one isn't stable unfortunately...
> 
> but `y` gets moved if not `Copy`, so it's not same

If you don't want to move you can do

```rust
x.as_ref() == Some(&y)
```

**Comment by Amanieu** [comment]
#108095 argues that this should be removed in favor of `is_some_with` (https://github.com/rust-lang/rust/issues/62358)) and I agree. That API is more flexible than `contains` and has a better name.

@rfcbot fcp close

**Comment by BrianLondon** [comment]
My anecdotal experience with Scala, which has an `Option[T]` with both `contains(t: T): bool` and `exists(t: T => bool): bool` the later of which is equivalent to the proposed `is_some_and`, is that use of `contains` is _far_ more common. I don't know enough about the compiler internals to speak to what performance penalty comes from wrapping the equality check in the closure, but if any, I'd be concerned about the added overhead.

**Comment by soc** [comment]
@BrianLondon Please refer to

> I wouldn't suggest throwing around your experience working on Scala as much as you do.

;

> If you just want to carbon copy the ideology of Scala into Rust, and are not interested in improving on the language, I suggest you go back and work on Scala

;

> Please keep in mind that, while your experience in Scala may be very good with applying Scala ideology to Scala problems, Rust might not fit in the same corner.

to learn about the community's preferences on this topic.

Thanks!

**Comment by BrianLondon** [comment]
Thanks @soc, is there a document you're quoting? If so can you link to the original please? I'd like to read the rest of it.

I think you're making some incorrect assumptions about my motivations. To be clear I am not suggesting a carbon copy of Scala, or even proposing to bring any particular Scala idiom into Rust. I am categorically not, for example, saying `is_some_and` should be renamed `exists` or something. What I am saying is that I expect `is_some_and(|x| x == target)` to be the predominant way it's used, and I'm asking whether there is a performance cost to running the comparison through the closure.

**Comment by camsteffen** [comment]
@BrianLondon, @soc is just quoting earlier comments in this thread where they got into a squabble with another individual, so no need to worry about that. Generally the community is welcoming to different perspectives.

**Comment by AaronKutch** [comment]
I personally think that `Option::contains` makes sense and should be a thing in parallel to `is_some_and`. `Result::contains`, however, will result in a cognitive ambiguity in `Result<T, T>` cases (I have used these multiple times in practice). For `Result` there should just be `is_ok_and` and `is_err_and`

**Comment by khionu** [comment]
As the moderator who handled the referred situation, I want to confirm that Brian's original comment from yesterday was not of the same problems. What was done well was that Brian noted that their experience was anecdotal, not treating their views as authoritative, and they also softened their contribution by acknowledging what they weren't aware of. This makes a world of difference in how one brings outside experience to Rust.

**Comment by soc** [comment]
@Centril methods have been removed, this issue can be closed now.

**Comment by soc** [comment]
Great outcome! Thanks to everyone who contributed in not getting this across the finish line for 3.5 years!

**Comment by bjorn** [comment]
@univerz As per above commit one could also replace by `iter().any(...)`, but it is only a little bit shorter so not any less annoying.

I'm really sad about the outcome of this issue. :-(

**Comment by soc** [comment]
@univerz Perhaps [option-ext](https://github.com/soc/option-ext) can be of use for you!
